sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1468,7 +1468,8 @@ object DecimalAggregates extends Rule[LogicalPlan] {
  */
 object ConvertToLocalRelation extends Rule[LogicalPlan] {
   def apply(plan: LogicalPlan): LogicalPlan = plan transform {
-    case Project(projectList, LocalRelation(output, data)) =>
+    case p @ Project(projectList, LocalRelation(output, data))
+        if !p.expressions.exists(ScalarSubquery.hasScalarSubquery) =>
```

Comment:
I think AttributeReference is the only exception, it will be replaced to BoundReference when create an Projection, we could have a special case for that.